incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -67,24 +95,50 @@ var _ = Describe("Secret", func() {
 		a := map[string]string{"propagate.hnc.x-k8s.io/select": "!" + bazName + ".tree.hnc.x-k8s.io/depth"}
 		makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
 
-		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue())
-		Consistently(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse())
+		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue(), "When propagating testrole to %s", barName)
+		Consistently(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse(), "When propagating testrole to %s", bazName)
+	})
+
+	It("should propagate object only to selected namespace using multiple select", func() {
+		setParent(ctx, barName, fooName)
+		setParent(ctx, bazName, fooName)
+
+		// Create a Role that does NOT propogate to bazName
+		a := map[string]string{"propagate.hnc.x-k8s.io/select": "!" + bazName + ".tree.hnc.x-k8s.io/depth, " +
+			"!" + barName + ".tree.hnc.x-k8s.io/depth"}
+		makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
+
+		Consistently(hasObject(ctx, "Role", barName, "testrole")).Should(BeFalse(), "When propagating testrole to %s", barName)
+		Consistently(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse(), "When propagating testrole to %s", bazName)
 	})
 
 	It("should remove object in selected namespace when selector is updated", func() {
 		setParent(ctx, barName, fooName)
 		setParent(ctx, bazName, fooName)
 
 		makeObject(ctx, "Role", fooName, "testrole")
-		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue())
-		Eventually(hasObject(ctx, "Role", bazName, "testrole")).Should(BeTrue())
+		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue(), "When propagating testrole to %s", barName)
+		Eventually(hasObject(ctx, "Role", bazName, "testrole")).Should(BeTrue(), "When propagating testrole to %s", bazName)
 
 		// Update the Role so that it does NOT propogate to bazName
 		a := map[string]string{"propagate.hnc.x-k8s.io/select": "!" + bazName + ".tree.hnc.x-k8s.io/depth"}
 		Expect(updateObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)).Should(Succeed())
 
-		Consistently(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue())
-		Eventually(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse())
+		Consistently(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue(), "When propagating testrole to %s", barName)
+		Eventually(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse(), "When propagating testrole to %s", bazName)
+	})
+
+	It("should propagate object only to selected namespace using both select and treeSelect when both are present", func() {
+		setParent(ctx, barName, fooName)
+		setParent(ctx, bazName, fooName)
```

Comment:
This would be a stronger test IMO again if there were three children. Use select to propagate to bar and baz; use treeSelect to propagate to baz and bam. Verify that it's only propagated to baz.

You might be able to set up a table test (as described above) that covers _all_ these cases. E.g.:

```
exceptionTests := []struct{
  name string
  select string
  treeSelect string
  want []string
}{
  {name: "simple treeSelect", treeSelect: barName, want: []string{barName}},
  {name: "negated treeSelect", treeSelect: "!"+barName, want: []string{bazName, bamName}},
  ... other simple one-line cases ...
  {
     name: "only propagate to namespaces selected by both select and treeSelect",
     select: barName+treeSuffix+","+bazName+treeSuffix,
     treeSelect: bazName+","+bamName,
     want: []string{bazName}
  },
  ... other complex cases ...
}
for _, tc := range exceptionTests {
  It("should handle exceptions: "+tc.name, func() {
    ... logic goes here ...
  }
}
```

Reply:
done